core/src/main/java/com/yahoo/oak/SyncRecycleMemoryManager.java

```diff
@@ -153,17 +153,17 @@ public void allocate(Slice s, int size, boolean existing) {
         //      initializing the header's lock to be free
         // for value being moved (existing): initialize the lock to be locked
         if (existing) {
-            HEADER.initLockedHeader(s, size);
+            HEADER.initLockedHeader(s.getMetadataAddress(), size, allocationVersion);
         } else {
-            HEADER.initHeader(s, size);
+            HEADER.initHeader(s.getMetadataAddress(), size, allocationVersion);
         }
-        assert HEADER.getVersion(s) == allocationVersion;
+        assert HEADER.getOffHeapVersion(s.getMetadataAddress()) == allocationVersion;
     }
 
     @Override
     public void release(Slice s) {
         if (s.length == Slice.UNDEFINED_LENGTH_OR_OFFSET) {
-            ValueUtilsImpl.setLengthFromOffHeap(s);
+            s.getLength(); // this will set the length from off-heap header, if needed
         }
```

Comment:
1. `s.getLength()` already checks `if (s.length == Slice.UNDEFINED_LENGTH_OR_OFFSET)`, so no need to check it here as well.
2. It is weird to find this piece of code here because the reader might not know about `Slice`'s lazy evaluation mechanism. Maybe add a method to `Slice`: `prefetchDataFromHeader()` that will fetch all data that is lazily stored on the header. This will allow you to put more data on the header in the future if needed.
Here you will call `s.prefetchDataFromHeader()` which will explicitly let the reader know of `Slice`'s lazy evaluation mechanism.

Reply:
fixed